test/e2e/deployment.go

```diff
@@ -1013,3 +1016,95 @@ func testScalePausedDeployment(f *framework.Framework) {
+	Expect(err).NotTo(HaveOccurred())
+
+	By(fmt.Sprintf("Waiting for deployment status to sync (current available: %d, minimum available: %d)", deployment.Status.AvailableReplicas, d.Spec.Replicas-2))
+	err = framework.WaitForDeploymentStatus(c, deployment, true)
+	Expect(err).NotTo(HaveOccurred())
+
+	first, err := deploymentutil.GetNewReplicaSet(deployment, c)
+	Expect(err).NotTo(HaveOccurred())
+
+	// Update the deployment with a non-existent image so that the new replica set will be blocked.
+	By(fmt.Sprintf("Updating deployment %q with a non-existent image", deploymentName))
+	deployment, err = framework.UpdateDeploymentWithRetries(c, ns, d.Name, func(update *extensions.Deployment) {
+		update.Spec.Template.Spec.Containers[0].Image = "nginx:404"
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	By(fmt.Sprintf("Waiting for observed generation %d", deployment.Generation))
+	err = framework.WaitForObservedDeployment(c, ns, deploymentName, deployment.Generation)
+	Expect(err).NotTo(HaveOccurred())
+
+	deployment, err = c.Extensions().Deployments(ns).Get(deploymentName)
+	Expect(err).NotTo(HaveOccurred())
+
+	By(fmt.Sprintf("Waiting for deployment status to sync (current available: %d, minimum available: %d)", deployment.Status.AvailableReplicas, d.Spec.Replicas-2))
+	err = framework.WaitForDeploymentStatus(c, deployment, false)
+	Expect(err).NotTo(HaveOccurred())
+
+	By(fmt.Sprintf("Checking that the replica sets for %q are synced", deploymentName))
+	second, err := deploymentutil.GetNewReplicaSet(deployment, c)
+	Expect(err).NotTo(HaveOccurred())
+
+	first, err = c.Extensions().ReplicaSets(first.Namespace).Get(first.Name)
+	Expect(err).NotTo(HaveOccurred())
+
+	firstCond := client.ReplicaSetHasDesiredReplicas(f.Client.Extensions(), first)
+	wait.PollImmediate(10*time.Millisecond, 1*time.Minute, firstCond)
+
+	secondCond := client.ReplicaSetHasDesiredReplicas(f.Client.Extensions(), second)
+	wait.PollImmediate(10*time.Millisecond, 1*time.Minute, secondCond)
+
+	By(fmt.Sprintf("Updating the size and template at the same time for deployment %q", deploymentName))
+	newReplicas := int32(20)
+	deployment, err = framework.UpdateDeploymentWithRetries(c, ns, deployment.Name, func(update *extensions.Deployment) {
+		update.Spec.Replicas = newReplicas
+		update.Spec.Template.Spec.Containers[0].Image = nautilusImage
+	})
+	Expect(err).NotTo(HaveOccurred())
+
+	err = framework.WaitForObservedDeployment(c, ns, deploymentName, deployment.Generation)
+	Expect(err).NotTo(HaveOccurred())
+
+	oldRSs, _, rs, err := deploymentutil.GetAllReplicaSets(deployment, c)
+	Expect(err).NotTo(HaveOccurred())
+
+	for _, rs := range append(oldRSs, rs) {
+		By(fmt.Sprintf("Ensuring replica set %q has the correct desiredReplicas annotation", rs.Name))
+		desired, ok := deploymentutil.GetDesiredReplicasAnnotation(rs)
+		if !ok || desired == deployment.Spec.Replicas {
+			continue
+		}
+		err = fmt.Errorf("unexpected desiredReplicas annotation %d for replica set %q", desired, rs.Name)
```

Comment:
Done

On Thu, Jul 21, 2016 at 12:17 AM, Phillip Wittrock <notifications@github.com
> wrote:

> In test/e2e/deployment.go
> <https://github.com/kubernetes/kubernetes/pull/28162#discussion_r71616830>
> :
>
> > +	})
> > +	Expect(err).NotTo(HaveOccurred())
> > +
> > +	err = framework.WaitForObservedDeployment(c, ns, deploymentName, deployment.Generation)
> > +	Expect(err).NotTo(HaveOccurred())
> > +
> > +	oldRSs, _, rs, err := deploymentutil.GetAllReplicaSets(deployment, c)
> > +	Expect(err).NotTo(HaveOccurred())
> > +
> > +	for _, rs := range append(oldRSs, rs) {
> > +		By(fmt.Sprintf("Ensuring replica set %q has the correct desiredReplicas annotation", rs.Name))
> > +		desired, ok := deploymentutil.GetDesiredReplicasAnnotation(rs)
> > +		if !ok || desired == deployment.Spec.Replicas {
> > +			continue
> > +		}
> > +		err = fmt.Errorf("unexpected desiredReplicas annotation %d for replica set %q", desired, rs.Name)
>
> This is scaling up + image update right? We also need scaling down + image
> update.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28162/files/b68dbfbffc725587ddfa28e82fa556f1b2deeaae#r71616830>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf3hnp4GfrcUubF3fnTQpkyvyag7Dks5qXp5kgaJpZM4JAOWg>
> .
>